WIP: Migrating OCaml binding to CMake

arbipher opened the pull request:
It's my WIP working on migrating the building of OCaml binding to CMake.

I think the majority of the work is done. I started by mimicking the other bindings and some llvm cmake for ocaml, but then I just manually wrote enough custom commands and dependencies. The actual building commands should have the same result as the previously generated python scripts, though I think I simplified some unnecessary dependencies and compiler flags.

tldr;
```console
  cd build && cmake \
-G "Ninja" \
-DZ3_BUILD_LIBZ3_SHARED=TRUE \
-DZ3_BUILD_OCAML_BINDINGS=TRUE \
-DZ3_USE_LIB_GMP=TRUE \
../
```

It should just work. I also build and run the `ml_examples.ml` for both bytecode and native code in the cmake to make sure it works. I observe all z3 bindings are using shared library so I wonder maybe I don't even need to make a static version for the ocaml-binding.

I also add one argument `-DZ3_BUILD_OCAML_EXTERNAL_LIBZ3=/path/to/libz3.so` to support building ocaml building with an external z3 library that is often installed by some other distro package managers.

Now the remaining work is :
1. The installation part. I really don't have a good idea for non opam installation.
2. Test it on macos. I knew the problem and the two-line fix for macos's problem. It's in the comments yet.
3. Remove unnecessary files.
4. Write some good document for it.
5. Tune for the CI.
6. (this is out of z3 but) Make opam package for this new building script. However, I guess I have to adjust both sides to try letting it work well.

p.s. for ocaml users, I also have a [post](https://discuss.ocaml.org/t/ask-for-suggestions-for-the-next-z3-package-release/8922/15) for discussion.

wintersteiger commented on cmake/modules/AddOCaml.cmake:
```diff
@@ -0,0 +1,269 @@
+    add_custom_command(TARGET "ocaml_${name}" POST_BUILD
+      COMMAND "${CMAKE_COMMAND}" "-E" "copy" "${install_file}"
+                                             "${LLVM_LIBRARY_DIR}/ocaml/llvm/"
+      COMMENT "Copying OCaml library component ${filename} to intermediate area"
```
```suggestion
      COMMENT "Installing OCaml library component ${filename}"
```

wintersteiger commented on README-CMake.md:
```diff
@@ -292,6 +292,9 @@ The following useful options can be passed to CMake whilst configuring.
 * ``Z3_JAVA_JNI_LIB_INSTALLDIRR`` - STRING. The path to directory to install the Z3 Java JNI bridge library. This path should be relative to ``CMAKE_INSTALL_PREFIX``.
+* ``Z3_BUILD_OCAML_BINDINGS`` - BOOL. If set to ``TRUE`` then Z3's OCaml bindings will be built.
+* ``Z3_BUILD_JULIA_BINDINGS`` - BOOL. If set to ``TRUE`` then Z3's Julia bindings will be built.
+* ``Z3_INSTALL_JULIA_BINDINGS`` - BOOL. If set to ``TRUE`` and ``Z3_BUILD_JULIA_BINDINGS`` is ``TRUE`` then running the ``install`` target will install Z3's Julia bindings.
```
Add similar option for the OCaml bindings?

wintersteiger reviewed (changes requested):
Overall this looks like a good start, but it needs a lot of clean-up and extensions before it will be usable in practice.

arbipher commented:
Thank you so much for your careful examination. This is one reason I would like to make this immature draft PR.
I am learning how z3 is installed on some platforms so my change won't break them. Also, I am still thinking how the ocaml-binding should be installed.

I will check your changes one by one later.

arbipher commented on makefile:
```diff
@@ -0,0 +1,400 @@
+	-ocamlcflags -bin-annot \
+	-package zarith \
+	-lz3 -lstdc++ -lpthread\
+	-lgmp \
```
Got it.

arbipher commented on src/api/ml/z3.mli:
```diff
@@ -3712,6 +3713,31 @@ end
     (set_global_param "pp.decimal" "true")
     will set the parameter "decimal" in the module "pp" to true.
+
+    Legal parameters are:
```
Oh, this comes from my previous learning/experimenting notes so it shouldn't be included here.

I agree with your point.

arbipher commented on src/api/ml/CMakeLists.txt:
```diff
@@ -0,0 +1,320 @@
+  set(libz3_path ${PROJECT_BINARY_DIR})
+endif()
+
+add_custom_command(
```
No. `add_custom_command` will check the `depends` and `outputs` to decide whether to run it. CMake also checks them statically to make sure these commands are valid.

arbipher commented on src/api/ml/CMakeLists.txt:
```diff
@@ -0,0 +1,320 @@
+  VERBATIM
+)
+
+add_custom_command(
```
It's subtle and I prefer to copy them. The reason is `ocamlc` may generate some intermediate files at the source location. It general, it can also simplify the dependencies of custom commands when some files are generated to be in the build path while some files are at the source path.

wintersteiger replied:
Oh yes, that's right. We do have to make sure everything is generated in the build directory, and sometimes that might indeed require copying source files around if the OCaml tools don't provide commandline flags to set the output path.

arbipher commented on src/api/ml/CMakeLists.txt:
```diff
@@ -0,0 +1,320 @@
+if( APPLE )
+  set(ocaml_rpath "@executable_path/../libz3${so_ext}")
+elseif( UNIX )
+  set(ocaml_rpath "\\$ORIGIN/../libz3${so_ext}")
```
At this moment, building on Windows/Cygwin is beyond my knowledge and I think even the support for OCaml on Windows is under heavy development. I may prefer to at least postpone it until the rest is done.

p.s. I notice there is some good news in the OCaml community.

wintersteiger replied:
Yeah, it's always been a bit neglected, but there are users, so we should make sure it works. BTW, the Opam people are also adding a Windows CI pipeline at the moment, so it might be required for Opam packages soon.

arbipher commented on src/api/ml/CMakeLists.txt:
```diff
@@ -0,0 +1,320 @@
+    ${z3ml_bin}/z3ml.cmxa
+    ${z3ml_bin}/dllz3ml.so
+    ${z3ml_bin}/ml_example.ml
+  COMMENT "Testing build and run ml_example natively"
```
Let me investigate it later. I have no idea about CMake-test, but if the major CMake for z3 uses it, I can change to that. This is more a sanity check to ensure the linking and building works, rather than ensure the same result or so.

arbipher commented on src/api/ml/CMakeLists.txt:
```diff
@@ -0,0 +1,320 @@
+# Install
+###############################################################################
+
+# Hacky: When the os is APPLE, a fix command will mutate `libz3.dylib` and `dlllibz3.so` inplace.
```
This is not so out-of-date but I will update it when macOS's solution for both building and installation is done. In the end, I may still leave some comments for macOS because it's kind of unlike Linux.

arbipher commented on cmake/modules/AddOCaml.cmake:
```diff
@@ -0,0 +1,269 @@
+# INCLUDES  Path for header files for C sources.
+# CFLAGS    Additional arguments passed when compiling C stubs.
+# PKG       Names of ocamlfind packages this library depends on.
+# LLVM      Names of LLVM libraries this library depends on.
```
It will remove it. It's the starting point of this work but later I realized the requirements for ocaml-llvm and ocaml-z3 are different.

NikolajBjorner commented:
It would be absolutely great to cmake-ify the ocaml bindings for uniform build support and allow to retarget release processes to use the more portable cmake system.

arbipher commented:
Aha, I was occupied by some random things after the previous effect on this issue in July. I hope to pick this up this month.

arbipher commented:
I made some updates (finally) to the PR. After it looks good to you, I can cherry-pick them to the latest commit and make another PR.

For the previous review, 
- I removed all the necessary source file copying.
- I deleted `AddOCaml.make`, which was used for llvm.

I added Github Action and made it work on both ubuntu and macOS. They [work](https://github.com/arbipher/z3/actions/runs/14551991533) (finally finally) on my forked repo.

For the linking especially rpath-related, since platforms like Debian (discourage), homebrew(hacking) or even opam have its own mechanism to handle, in this build I just figured out a way (mainly for macOS) to make two working build and run as sanity check. 

The next step is to allow external libz3 and allow binding to be built with them. I am still not very clear on the current Windows support for ocaml as well as the users will use a Windows native binding.

NikolajBjorner commented:
Regarding Windows + OCaml: It isn't (was never before) a big use case and nowadays it is very easy to start up wsl2/ubuntu for ocaml development.

I sometimes tease @msprotz on his past heroic efforts in this space, so thanks for the reminder.

arbipher commented:
Cool for the merging. There is a debug helper `makefile` in it (which shouldn't be a big deal and I will remove it in the next PR for the next steps).
